rust/syntect-plugin/src/main.rs

```diff
@@ -82,35 +71,56 @@ impl<'a> PluginState<'a> {
         }
         let line = line.unwrap();
         let ops = self.parse_state.as_mut().unwrap().parse_line(&line);
-        if self.builder.is_none() {
+        if self.spans.is_empty() {
             self.spans_start = self.offset;
-            self.builder = Some(SpansBuilder::new());
         }
-        let iter = HighlightIterator::new(self.hstate.as_mut().unwrap(), &ops, &line,
-            self.highlighter.as_ref().unwrap());
-        let mut ix = 0;
-        for (style, str_slice) in iter {
-            let start = self.offset - self.spans_start + ix;
-            let end = start + str_slice.len();
-            add_style_span(self.builder.as_mut().unwrap(), style, start, end);
-            ix += str_slice.len();
+
+        //print_err!("\n\nline {}\napplying ops {:?}", &line, &ops);
+        let mut prev_cursor = 0;
+        for (cursor, batch) in ops {
+            if self.scope_state.len() > 0 {
+                let scope_ident = self.stack_idents.get_value(self.scope_state.as_slice());
+                //print_err!("scope ident: {:?}", scope_ident);
+                let scope_ident = match scope_ident {
+                    LookupResult::Existing(id) => id,
+                    LookupResult::New(id) => {
+                        let stack_strings = self.scope_state.as_slice().iter()
+                            .map(|slice| slice.build_string())
```

Comment:
So, this will probably be a bottleneck, because this operation isn't meant to be efficient when called like this.

All scopes in `syntect` store their actual string contents in a shared atom repository called `SCOPE_REPO` that this method locks each time it is called, which is terribly inefficient.

Instead what you can do is lock `SCOPE_REPO` (it's public) once outside this loop and use its `to_string` method here. See [here](https://github.com/trishume/syntect/compare/unique-scopes#diff-35cf218b39de628b31722bb466ecb837R84) for an example of how to do this.